src/init.cpp

```diff
@@ -1149,6 +1150,21 @@ bool AppInitParameterInteraction()
             }
         }
     }
+
+    if (gArgs.IsArgSet("-segwitheight")) {
+        if (!chainparams.MineBlocksOnDemand()) {
+            return InitError("Segwit activation parameters may only be overridden on regtest.");
+        }
+        int64_t height = gArgs.GetArg("-segwitheight", chainparams.GetConsensus().SegwitHeight);
+        if (height < -1 || height >= std::numeric_limits<int>::max()) {
+            return InitError(strprintf("Activation height %ld for segwit is out of valid range. Use -1 to disable segwit.", height));
+        }
+        else if (height == -1) {
+            LogPrintf("Segwit disabled for testing\n");
```

Comment:
It seems odd to add code just for the tests, that has no application outside of regtest. I'd prefer if the code paths in the tests that test segwit can be disabled are just removed. Afterward this code block and all other dead code (`IsScriptWitnessEnabled`, ...) can be discarded.

Reply:
I completely agree. The only reason the 'Segwit disabled' code here and elsewhere exists is for `p2p_segwit.py`, where it's used to test whether GBT doesn't return a default witness commitment if segwit is disabled. That's completely circular - we're adding behaviour into the product so we can test whether that behaviour exists.

I'd like to completely overhaul p2p_segwit.py and remove this special casing, but in a future PR. This PR is simply maintaining the existing functionality of setting `-vbparams=segwit:0:0`.